ocaml/xapi/xapi_pool_helpers.ml

```diff
@@ -269,3 +269,308 @@ let apply_guest_agent_config ~__context =
+        in
+        let execute_next_month =
+          if not first_run then
+            true
+          else if
+            monthday_this_month < tm_now.Unix.tm_mday
+            || monthday_this_month = tm_now.Unix.tm_mday
+               && hour_configed_int <= tm_now.Unix.tm_hour
+          then
+            true
+          else
+            false
+        in
+        let month_next_run, year_next_run =
+          if execute_next_month then
+            if tm_now.Unix.tm_mon = 12 then
+              (1, tm_now.Unix.tm_year + 1)
+            else
+              (tm_now.Unix.tm_mon + 1, tm_now.Unix.tm_year)
+          else
+            (tm_now.Unix.tm_mon, tm_now.Unix.tm_year)
+        in
+        let monthday_next_run =
+          if day_configed_int > 0 then
+            day_configed_int
+          else if not execute_next_month then
+            monthday_this_month
+          else
+            let year_2_months_later, month_next_next =
+              if tm_now.Unix.tm_mon = 12 then
+                (tm_now.Unix.tm_year + 1, 2)
+              else if tm_now.Unix.tm_mon = 11 then
+                (tm_now.Unix.tm_year + 1, 1)
+              else
+                (tm_now.Unix.tm_year, tm_now.Unix.tm_mon + 2)
+            in
+
+            let smallest_secs_2_months_later, _ =
+              Unix.mktime
+                {
+                  tm_year= year_2_months_later
+                ; tm_mon= month_next_next
+                ; tm_mday= 1
+                ; tm_hour= 0
+                ; tm_min= 0
+                ; tm_sec= 0
+                ; tm_wday= 0
+                ; tm_yday= 0
+                ; tm_isdst= false
+                }
+            in
+
+            let secs_configed =
+              Float.to_int smallest_secs_2_months_later
+              + (day_configed_int * 24 * 60 * 60)
+              + (hour_configed_int * 60 * 60)
+            in
+            let tm_configed = Unix.gmtime (Int.to_float secs_configed) in
+            tm_configed.Unix.tm_mday
+        in
+        let min_random = secs_random_within_an_hour / 60 in
+        let sec_random = secs_random_within_an_hour mod 60 in
+        let secs_next_run, _ =
+          Unix.mktime
+            {
+              tm_year= year_next_run
+            ; tm_mon= month_next_run
+            ; tm_mday= monthday_next_run
+            ; tm_hour= hour_configed_int
+            ; tm_min= min_random
+            ; tm_sec= sec_random
+            ; tm_wday= 0
+            ; tm_yday= 0
+            ; tm_isdst= false
+            }
+        in
+        debug
+          "[PeriodicUpdateSync] periodic_update_sync_delay monthly, \
+           execute_next_month=%B year_next_run=%d month_next_run=%d \
+           monthday_next_run=%d"
+          execute_next_month (year_next_run + 1900) (month_next_run + 1)
+          monthday_next_run ;
+        secs_next_run -. secs_now
+
+  let update_sync_delay ~__context ~first_run =
+    if Xapi_fist.update_sync_every_ten_minutes () then (
```

Comment:
The fist should go through the major function `periodic_update_sync_delay`. Otherwise, the main changes will be ignored by the testings which leverage fist.